Use Arc features in datasource extensions for eager startup and active/inactive

yrodiere opened the pull request:
~~Based on #41810 , which must be merged first.~~ => Done

Best reviewed commit by commit.

Effects of this PR:

1. Fixes #36666: see changes in expected behavior in `*DatasourceHealthCheckTest.java` for Agroal.
2. Fixes #43517:  see changes in expected behavior in `*DatasourceHealthCheckTest.java` and `*ConfigUrlMissing*Test.java` for reactive SQL client extensions.
3. Improves error messages when an application attempts to statically inject or dynamically access a Datasource bean that is deactivated and/or has no URL set.
4. Introduces a few (arguably beneficial) breaking changes, see below.

Content to be added to the migration guide:

`````asciidoc

== Datasources

[[datasource-no-default-url]]
=== No default URL for Reactive SQL clients

In previous versions of Quarkus, when dev services were disabled, or when running in production mode, `quarkus.datasource.reactive.url`/`quarkus.datasource."datasource-name".reactive.url`  would be implicitly set to an undocumented default targeting `localhost` and a DB-specific port.

Starting with this version of Quarkus, when dev services are disabled, or when running in production mode, `quarkus.datasource.reactive.url`/`quarkus.datasource."datasource-name".reactive.url` no longer has a default: if the property is not set, the corresponding datasource will be deactivated, leading to the behavior described in <<datasource-inactive-fail-fast>>.

If your application needs to target localhost for one of its datasources, set `quarkus.datasource.reactive.url`/`quarkus.datasource."datasource-name".reactive.url` explicitly, for example in `application.properties`.

[[datasource-no-health-check-if-no-url]]
=== Datasources without a URL no longer contribute a health check

In previous versions of Quarkus, when dev services were disabled, or when running in production mode, and when `quarkus.datasource.reactive.url`/`quarkus.datasource."datasource-name".reactive.url` was not set, a health check would be contributed for the corresponding datasource regardless, resulting in unreliable health checks that would either always succeed (JDBC datasource) or (almost) always fail (reactive datasources).

Starting with this version of Quarkus, datasources without a URL will no longer contribute a health check.

If your application needs a health check for a datasource, make sure to set `quarkus.datasource.jdbc.url`/`quarkus.datasource."datasource-name".jdbc.url`/`quarkus.datasource.reactive.url`/`quarkus.datasource."datasource-name".reactive.url` explicitly.

NOTE: Quarkus will still attempt to detect misconfiguration on application startup, but will report it differently, potentially earlier: see <<datasource-inactive-fail-fast>>.

[[datasource-inactive-fail-fast]]
=== Datasource usage fails fast if datasource is deactivated or has no URL set

In previous versions of Quarkus, when dev services were disabled, or when running in production mode, and when the datasource was deactivated (`quarkus.datasource.active=false`) or had no URL set, the application would be allowed to start successfully, but could fail much later upon first access to the datasource.

Starting with this version of Quarkus, datasources that are deactivated or do not have a URL will lead to a startup failure if Quarkus can detect they are used:

* Static CDI injection points involving the datasource (`@Inject DataSource ds` or `@Inject Pool pool`) will cause application startup to fail with an explicit, actionable message.
* Dynamic retrieval of the datasource (e.g. through `CDI.getBeanContainer()`/`Arc.instance()`, or by injecting an `Instance<DataSource>`) will not be detected on startup, but will cause an exception to be thrown with an explicit, actionable message.

The <<flyway-inactive-fail-fast,Flyway>> and <<liquibase-inactive-fail-fast,Liquibase>> extensions implement a similar behavior for their `Flyway`/`LiquibaseFactory` CDI beans.

If your application needs to inject a `DataSource`/`Flyway`/`LiquibaseFactory` bean for a datasource that can potentially be deactivated or have no URL, you may encounter a startup failure similar to this:

```
io.quarkus.arc.InactiveBeanException: Bean is not active: SYNTHETIC bean [class=io.agroal.api.AgroalDataSource, id=sqqLi56D50iCdXmOjyjPSAxbLu0]
Reason: Datasource '<default>' was deactivated automatically because its URL is not set. To activate the datasource, set configuration property 'quarkus.datasource.jdbc.url'. Refer to https://quarkus.io/guides/datasource for guidance.
To avoid this exception while keeping the bean inactive:
	- Configure all extensions consuming this bean as inactive as well, if they allow it, e.g. 'quarkus.someextension.active=false'
	- Make sure that custom code only accesses this bean if it is active
	- Inject the bean with 'Instance<io.agroal.api.AgroalDataSource>' instead of 'io.agroal.api.AgroalDataSource'
This bean is injected into:
	-io.quarkus.agroal.test.ConfigUrlMissingDefaultDatasourceStaticInjectionTest$MyBean#ds
	at io.agroal.api.AgroalDataSource_sqqLi56D50iCdXmOjyjPSAxbLu0_Synthetic_Bean.doCreate(Unknown Source)
	[...]
```

To avoid this failure, follow the instructions in the exception message.
In particular, consider injecting the problematic bean as an `InjectableInstance` (for example `@Inject InjectableInstance<DataSource> ds`) instead.
You can check whether that bean is active using `.getHandle().getBean().isActive().result()` on the `InjectableInstance`, and retrieve the bean instance using `InjectableInstance#get()`.
See https://quarkus.io/guides/datasource#datasource-active for an example of such a setup for `DataSource` beans -- the idea would be similar for a `Flyway`/`LiquibaseFactory` beans.

== Flyway

[[flyway-inactive-inactive-fail-fast]]
=== Flyway usage fails fast if datasource is deactivated or has no URL set

Starting with this version of Quarkus, datasources that are deactivated or have no URL set will lead to a startup failure if Quarkus can detect a `Flyway` bean is used for that datasource.

See the <<datasource-inactive-fail-fast,datasources,similar migration guide entry for datasources>> for more information.

== Liquibase

[[liquibase-inactive-if-no-url]]
=== Liquibase usage fails fast if datasource is deactivated or has no URL set

Starting with this version of Quarkus, datasources that are deactivated or have no URL set will lead to a startup failure if Quarkus can detect a `LiquibaseFactory` bean is used for that datasource.

See the <<datasource-inactive-fail-fast,datasources,similar migration guide entry for datasources>> for more information.
`````

yrodiere commented:
Actually I'll keep this as a draft... the base commit makes changes in Arc, which probably means the rebuild will be massive, even if I only need to test things related to datasources :/

yrodiere commented:
Hey @Ladicek, FYI I made some progress here regarding integration with your work in #41810, but didn't find the time to finish before my PTO. Sorry... I'll resume my work when I'm back at the end of August.

I'd personally not bother merging #41810 until then, but feel free if you're in a rush and are okay with what you're seeing here so far. I guess we can always iterate later on.

See you!

Ladicek commented:
@yrodiere Thanks, and no worries -- I have intentionally made #41810 a draft and there's no need to rush it in. I'll wait for more of your feedback, and maybe change the API a little bit before making it ready, we'll see. Enjoy your PTO!

yrodiere committed:
Fail on startup when inactive datasources are injected into user beans

By reimplementing Datasource inactive/active handling and eager startup
through Arc's native features, which is better integrated and gives us
this behavior.

yrodiere committed:
Always trim datasource URLs

So that we correctly detect empty URLs.

yrodiere committed:
Remove NoConfigTest for datasources

Because it duplicates:

* ConfigUrlMissingDefaultDatasourceStaticInjectionTest
* ConfigUrlMissingDefaultDatasourceDynamicInjectionTest
* ConfigUrlMissingNamedDatasourceStaticInjectionTest
* ConfigUrlMissingNamedDatasourceDynamicInjectionTest

yrodiere committed:
Fix incorrect testing conditions in "active-false" tests for reactive datasources

yrodiere committed:
Fail on startup when Flyway/Liquibase for inactive datasources are injected into user beans

By reimplementing Flyway's/Liquibase's inactive/active handling and eager
startup through Arc's native features, which is better integrated and gives
us this behavior.

yrodiere committed:
Update Hibernate ORM's "active" documentation

For consistency with datasource documentation.

yrodiere committed:
Do not rely on default URL of reactive datasources in tests

yrodiere commented:
Tests are passing on my fork: https://github.com/yrodiere/quarkus/actions/runs/11109607551

Let's wait for #41810 to get merged, then I'll rebase and ask somone for a review.

yrodiere committed:
Deactivate datasources when their URL is unset at runtime

Consequently:

* Skip the health check for datasources without a URL
* Fail on startup if datasources without URL are (statically)
  injected into user bean
* Fail on first access if datasources without URL are retrieved
  dynamically through CDI

tsegismont commented on extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/DataSourceReactiveRuntimeConfig.java:
```diff
@@ -31,7 +33,7 @@ public interface DataSourceReactiveRuntimeConfig {
      * Note that certain drivers might not accommodate multiple values in this context.
      */
-    Optional<List<String>> url();
+    Optional<List<@WithConverter(TrimmedStringConverter.class) String>> url();
```
I wasn't aware of this trick, thanks

tsegismont commented on extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/ReactiveDataSourceUtil.java:
```diff
@@ -13,6 +18,42 @@ public final class ReactiveDataSourceUtil {
+            var bean = handle.getBean();
+            if (bean != null && bean.isActive()) {
+                String name = dataSourceName(bean);
+                if (name != null) { // There may be custom beans, these will have a null name and will be ignored.
```
:+1:

tsegismont commented on extensions/reactive-db2-client/runtime/src/main/java/io/quarkus/reactive/db2/client/runtime/DB2PoolRecorder.java:
```diff
@@ -50,6 +52,33 @@ public class DB2PoolRecorder {
     private static final TypeLiteral<Instance<DB2PoolCreator>> POOL_CREATOR_TYPE_LITERAL = new TypeLiteral<>() {
     };
 
+    private final RuntimeValue<DataSourcesRuntimeConfig> runtimeConfig;
+    private final RuntimeValue<DataSourcesReactiveRuntimeConfig> reactiveRuntimeConfig;
```
IIUC a new recorder instance is created for each build step invocation?

yrodiere replied:
I'm not sure, but at static/runtime init I'd expect one recorder instance per build step, not one per call made from a build step.

Why this question though?

tsegismont reviewed (approved):
Looks good to me, thank you @yrodiere

marko-bekhta commented on docs/src/main/asciidoc/datasource.adoc:
```diff
@@ -445,22 +445,26 @@ AgroalDataSource inventoryDataSource;
 === Activate or deactivate datasources
 
-When a datasource is configured at build time, it is active by default at runtime.
+When a datasource is configured at build time, it is active at runtime as soon as its URL is set.
```
maybe?
```suggestion
When a datasource is configured at build time and its URL is set at runtime, it is active by default.
```
it feels like if we say `as soon as` it's more about the timing of activating the datasource rather than the fact itself, but IDK 🙈 🤷🏻‍♂️

marko-bekhta commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:
```diff
@@ -56,28 +56,42 @@ public void setApplicationCallbackClasses(Map<String, Collection<Callback>> call
-                            dataSourceName, e.getMessage()), e);
-                }
-
+                // ClientProxy.unwrap is necessary for the instanceof check below.
```
Is it still the case? Looks like the instanceof checks are gone, and I didn't notice any apparent reason looking into `flywayProducer.createFlyway(..)`.

Looking more at it ... I see that in a similar place in the liqubase extension, this unwrapping is gone

marko-bekhta reviewed (commented):
Looks nice! 🎉

gastaldi commented on extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DataSourceRuntimeConfig.java:
```diff
@@ -15,15 +16,11 @@ public interface DataSourceRuntimeConfig {
      * @asciidoclet
      */
     @WithDefault("true")
-    boolean active();
+    @ConfigDocDefault("`true` if the URL is set, `false` otherwise")
+    Optional<Boolean> active();
```
If there is a `WithDefault`, it doesn't make sense to use `Optional`, good catch!

yrodiere replied:
Keeping the default=true was an oversight.

I wanted to use `Optional<Boolean>` so that we can detect misuse; for example in Hibernate ORM you can set `quarkus.hibernate-orm.enabled=false` and `quarkus.hibernate-orm.active=true`, both explicitly, and it will result in an error because, well, that doesn't make any sense.

Here, I'm wondering if we should raise an error when setting `quarkus.datasource.active=true` explicitly, but not setting a URL... Though you might want to make the datasource active, but only set the JDBC URL.

Oh, well, I'll just drop the optional, it's a problem for another day.

yrodiere replied:
> Oh, well, I'll just drop the optional, it's a problem for another day.

yrodiere replied:
> Oh, well, I'll just drop the optional, it's a problem for another day.

Actually, you could say the datasource isn't really active if the URL is not set, so... I'll drop `@WithDefault`, so that the next person using this config will at least ask themselves "what the hell does it mean when it's absent?".

yrodiere commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalRecorder.java:
```diff
@@ -20,14 +36,32 @@ public AgroalDataSourceSupport get() {
+            @Override
+            public ActiveResult get() {
+                Optional<Boolean> active = runtimeConfig.getValue().dataSources().get(dataSourceName).active();
+                if (active.isPresent() && !active.get()) {
```
I agree my code is not great, but I personally find yours even more confusing... A matter of tastes, I guess :grinning:

geoand commented:
> @geoand (or @andrejpetras ?) could you please have a look at the impact on Liquibase?

Seems reasonable

yrodiere committed:
Simplify tests involving multiple datasources as alternatives

yrodiere committed:
Improve documentation for multiple-datasource-alternative setups

Mention the "getActive()" solution in particular.

yrodiere commented:
I rebased on main and added documentation for the `getActive()` solution implemented by @Ladicek in #43803 .

Will merge in the next few days unless someone objects.

geoand reviewed (commented):
Rubber-stamped!

yrodiere commented:
Merged! Thanks a lot to everyone involved :)

yrodiere commented:
Also, I updated the 3.17 migration guide: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.17#datasources

geoand commented:
👍🏽

michalvavrik commented on extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolRecorder.java:
```diff
@@ -49,6 +51,33 @@ public class PgPoolRecorder {
+    private final RuntimeValue<DataSourcesRuntimeConfig> runtimeConfig;
+    private final RuntimeValue<DataSourcesReactiveRuntimeConfig> reactiveRuntimeConfig;
+
+    @Inject
```
interesting, does `jakarta.inject.Inject;` has any effect on records at all? I didn't know.

Ladicek replied:
It doesn't really. If the annotation wasn't present, it would work exactly the same.

If anything, this is more of a documentation that injection happens. But it's way out of any specification, so perhaps it shouldn't be here after all :shrug:

michalvavrik replied:
copy that, thanks
